### PR TITLE
WIP Add complete unit tests to the weight_limit spec

### PR DESCRIPTION
**Pull request by fredericboivin**
The weight limit spec and the associated methods contains a lot of code who looks unreachable or that we're not sure why or how it works (or even if we need it)

The plan would be to remove, refactor or extract some of those parts but they are virtually untested so I've begun adding unit tests for each individual methods.

This is a work in progress, comments and suggestions are very welcome.

I also used transpec to update the syntax.

### CODE REVIEW

File: spec/models/weight_limits_spec.rb [hunk 1]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
       Spree::ActiveShipping::Config.set(:default_weight => 1)
     end
 
+    describe ".valid_weight_for_package?" do
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:104]: Usually we do not test private methods, they should be tested as part of the behavior of the public method that they are called from. In this particular case the `valid_weight_for_package?` helper method should be tested as part of the expected behavior for the `ActiveShipping::Base#available?`. The tests for that should probably live in a `spec/models/spree/calculator/shipping/active_shipping/base_spec.rb` file however. Currently there are some tests for that in `spec/models/active_shipping_calculator_spec.rb` but we should probably move those in the correct place.

File: spec/models/weight_limits_spec.rb [hunk 2]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+      end
+
+      context "with a nil max weight" do
+        it "should return false if the max_weight is nil" do
+          expect(domestic_calculator.send :valid_weight_for_package?, package, nil).to eq false
+        end
+      end
+    end
+
+    describe ".country_weight_error?" do
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:133]: Same as above, this should be tested as part of the tests for `ActiveShipping::Base#available?`.

File: spec/models/weight_limits_spec.rb [hunk 3]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+        active_shipping_weights = [variant1.weight, variant2.weight, variant3.weight].map do |x|
+            (x * Spree::ActiveShipping::Config[:unit_multiplier]).to_d
+        end
+        expect(weights).to match_array active_shipping_weights
+      end
+
+      context "with a package containing variants with no weight" do
+        before do
+          # It's not set globally, therefore we set a value before those tests
+          Spree::ActiveShipping::Config.set(:max_weight_per_package => 30)
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:162]: We should make sure that we change the config in `around` blocks and set the value back to the default after the tests dependent on the configuration are run. We had issues with this leaking into other tests and creating test-order dependent failures.

File: spec/models/weight_limits_spec.rb [hunk 4]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+          end
+
+          expect(weights).to match_array active_shipping_weights
+        end
+      end
+
+      context "with a package containing variants with individual item weight less than or equal to the maximum per package weight" do
+        before do
+          # It's not set globally, therefore we set a value before those tests
+          Spree::ActiveShipping::Config.set(:max_weight_per_package => 100)
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:180]: Same as above, this should be set in an `around` block and re-set after examples are run.

File: spec/models/weight_limits_spec.rb [hunk 5]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+            (x * Spree::ActiveShipping::Config[:unit_multiplier]).to_d
+          end
+          expect(weights).to match_array active_shipping_weights
+        end
+      end
+
+      context "with a package containing variants with an individual item weight higher than the maximum per package weight" do
+        before do
+          # It's not set globally, therefore we set a value before those tests
+          Spree::ActiveShipping::Config.set(:max_weight_per_package => 1)
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:195]: Same as above.

File: spec/models/weight_limits_spec.rb [hunk 6]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+
+        it "should raise a Spree::ShippingError" do
+          expect{domestic_calculator.send :convert_package_to_weights_array, small_package }.to raise_error(Spree::ShippingError)
+        end
+      end
+
+      context "with a package containing variants where total weight (item weight * quantity) is higher than the maximum per package weight" do
+        before do
+          # It's not set globally, therefore we set a value before those tests
+          Spree::ActiveShipping::Config[:unit_multiplier] = 1
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:206]: Same as above.

File: spec/models/weight_limits_spec.rb [hunk 7]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+        it "should return an array with a value for each individual item (1 product x 2 counts as two separate items)" do
+          packages = domestic_calculator.send :convert_package_to_weights_array, package_single_item
+          expect(packages.size).to eq 2
+        end
+      end
+
+      context "with a package containing variants where total weight (item weight * quantity) is less than or equal to the maximum per package weight" do
+        before do
+          # It's not set globally, therefore we set a value before those tests
+          Spree::ActiveShipping::Config[:unit_multiplier] = 1
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:226]: Same as above.

File: spec/models/weight_limits_spec.rb [hunk 8]
```diff
@@ -91,6 +101,171 @@ def build_inventory_unit(variant, order)
...
+
+      context "with a package within the weights limit" do
+         it "should return false" do
+          allow(domestic_calculator).to receive(:valid_weight_for_package?).and_return(true)
+          expect(domestic_calculator.send :country_weight_error?, package).to be_nil
+        end
+      end
+    end
+
+    describe ".convert_package_to_weights_array" do
```
Comment (forkata) [inline spec/models/weight_limits_spec.rb:149]: This should be tested as part of the tests for `ActiveShipping::Base#compute_package`. Having said that testing all this as part of `#compute_package` will probably be quite messy, so I would be in favor of extracting the packages and weights calculation to a separate class, as it isn't the calculators job to deal with all this. I know that the goal is to have tests for everything before we start refactoring, but in this particular case I think extracting the behavior and adding the tests in the same step may be easier.

File: spec/models/weight_limits_spec.rb (hunk 3 above)
Comment (fredericboivin) [inline spec/models/weight_limits_spec.rb:162]: You're right. Good catch

File: spec/models/weight_limits_spec.rb (hunk 1 above)
Comment (fredericboivin) [inline spec/models/weight_limits_spec.rb:104]: (Adressing both comments) 

I understand your point, though I have some concerns : there are multiple levels of indirection between all the functions being called and the behavior is not consistent (sometimes it's rescued, sometimes it's not ; some valid paths are crashing because of bugs in the code related to nil weight or max weight limit, some paths are not used at all), so it is hard to test all the possible edge cases from the compute or similar higher level methods, and testing each of those invidually is helpful in finding the underlying bugs.

Maybe those tests does not need to be merged but I think they are really useful in understanding the messy parts of the code and documenting current behavior before it can be extracted.

As for the location of the tests file, you're also right. I didn't concern myself with it too much because I was planning on merging the two files once I was finished, and the namespace should be corrected.

Maybe I can just extract it into a new class for now without changing any of the code? It will be still be easier to test and both specs will be leaner and more concise.

Either way, thanks for the feedback.

Comment (forkata) [inline spec/models/weight_limits_spec.rb:104]: I agree with you on all of this. I think if we can extract all of the package weight stuff into a class without modifying the logic and adding the tests you have for that. We'll be in a much better place and avoid testing these private methods.

Let me know if you want to chat more about what that might look like. Thanks again for going through and figuring out all this convoluted logic :heart:

**Comment by forkata** [comment]
@dangerdogz If you want, pull the first commit in a separate PR and we can run transpec on all the specs and merge that separately while you are working on the weight refactor/specs.

File: spec/models/weight_limits_spec.rb (hunk 1 above)
Comment (fredericboivin) [inline spec/models/weight_limits_spec.rb:104]: Working on it !

**Comment by fredericboivin** [comment]
Won't be used since we went another way but I documented some bugs in there so leaving it around until I can fix them for sure (related to issue #28 )
